pkg/filesystem/pool/sector_pointer.go

```diff
@@ -0,0 +1,461 @@
+package pool
+
+import (
+	"io"
+	"math"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// NewSectorPointer returns a new SectorPointer struct.
+func NewSectorPointer() *SectorPointer {
+	return &SectorPointer{}
+}
+
+// SectorPointer is a struct which implements mapping from logical to
+// physical sectors in a manner based on the unix inode pointer
+// structure.
+//
+// It is a compact struct designed to densely represent the sectors used
+// by a file. Since the vast majority of files are small and they tend
+// to cluster their data this struct is a good fit for our use cases.
+//
+// A value of 0 in a direct array (or a nil pointer in an indirect
+// array) means that there is no sector allocated for that index.
+// Otherwise the value corresponds to the sector index where the data
+// can be found.
+//
+// The larger the file the more indirections it allows. For a typical
+// 4096 byte sector size this results in:
+//   - No indirection for files up to ~40KiB
+//   - Single indirection for files up to ~4MiB
+//   - Double indirection for files up to ~4GiB
+//   - Triple indirection for files up to ~4TiB
+type SectorPointer struct {
+	direct              [12]uint32
+	single              *indirect[uint32]
+	double              *indirect[*indirect[uint32]]
+	triple              *indirect[*indirect[*indirect[uint32]]]
+	logicalSectorLength uint32
+}
+
+// indirectionArraySize is the base size of the indirection array.
+//
+// directSectors, singleIndirectionSectors, doubleIndirectionSectors,
+// and tripleIndirectionSectors are the maximum number of sectors which
+// can be represented by a maximum of 0, 1, 2, and 3 indirections
+// respectively.
+const (
+	indirectionArraySize     = 1626 // Smallest size where the full unsigned 32 bit range can be addressed.
+	directSectors            = 12
+	singleIndirectionSectors = directSectors + indirectionArraySize
+	doubleIndirectionSectors = singleIndirectionSectors + indirectionArraySize*indirectionArraySize
+	tripleIndirectionSectors = doubleIndirectionSectors + indirectionArraySize*indirectionArraySize*indirectionArraySize
```

Comment:
Can you please put this underneath?
```go
_ uint = tripleIndirectionSectors - 1<<32
```
This will throw a compile-time error if `indirectionArraySize` is too small.